### PR TITLE
Exposing multi highlight and selection feature to js api

### PR DESCRIPTION
**Pull request by samcake**
This pr expose the selection and highlighting features from the JS api.

We did one change to the addToSelectedItemsList to avoid to add the same objects multiple times, an object is added only once to a selection list.

A script can add and remove object(s) from a Selection List and explore its content using the Selection api functions:
- addToSelectedItemsList(listName, itemType, ID)
- removeFromSelectedItemsList(listName, itemType, ID)
- clearSelectedItemsList(listName)
- getSelectedItemsList(listName)
- removeListFromMap(listName)
- getListNames()

The objects in a list can then be displayed highlighted by enabling the feature and configuring it through the api functions:
- enableListHighlight(listName, highlightStyle)
- disableListHighlight(listName)
- getListHighlightStyle(listName)
- getHighlightedListNames()

The highlight effect can display an outline and a filling color and opacity drawn over the main 3d view. The style offer the following configuration:
- outlineUnoccludedColor: {xColor} Color of the specified highlight region
- outlineOccludedColor: {xColor} "
- fillUnoccludedColor: {xColor} "
- fillOccludedColor: {xColor} "
- outlineUnoccludedOpacity: {float} Opacity value ranging from 0.0 (not visible) to 1.0 (fully opaque) for the specified highlight region
- outlineOccludedOpacity: {float}
- fillUnoccludedOpacity: {float}
- fillOccludedOpacity: {float} 
- outlineWidth: {float} width of the outline expressed in pixels
- isOutlineSmooth: {bool} true to enable oultine smooth falloff

We also added a revised version of the tool script "debugHighlight.js" which exposeillustrate how to use the Selection & Highlight api while allowing to inspect the values for the current highlighted selection defined.  

## TEST PLAN
1/ Run this test:  https://github.com/highfidelity/hifi_tests/blob/master/tests/engine/render/effect/highlight/coverage/test.md

2/ run the script "debugHighlight.js" and fool around with the fields driving the highlight style

3/ Test the highlighting feature used in the "Overlay context info" displayed when right clicking on an object from the marketplace

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5516/HighFidelity-Beta-PR11952-013d2e808d8025bfac8df4e27fd714b64f29747c.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5520/HighFidelity-Beta-PR11952-9a4f66705971a681b3eff4448997430f1d13c51e.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5561/HighFidelity-Beta-PR11952-429d54d65d4404fc283cca64a0f6efcbb5ad4322.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5562/HighFidelity-Beta-PR11952-eec4e966a75084d17cf684fda24a25f2246a7875.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5562/HighFidelity-Beta-PR11952-eec4e966a75084d17cf684fda24a25f2246a7875.dmg

**Comment by ctrlaltdavid** [comment]
Should "opacity" be "alpha", in line with entity and overlay properties?

**Comment by samcake** [comment]
@ctrlaltdavid Yep, sounds good to me, it s better if its unifying the names.
i ll check that up

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5564/HighFidelity-Beta-PR11952-bb888f2d4ce7cc6d7f271d48fb562851013602cd.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5564/HighFidelity-Beta-PR11952-bb888f2d4ce7cc6d7f271d48fb562851013602cd.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5565/HighFidelity-Beta-PR11952-cc4b2666478a9c63f6793efa1f2137cb1150132a.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5565/HighFidelity-Beta-PR11952-cc4b2666478a9c63f6793efa1f2137cb1150132a.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5566/HighFidelity-Beta-PR11952-d1cd9f5c6b1661e63c719b71263c4ab719492762.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5566/HighFidelity-Beta-PR11952-d1cd9f5c6b1661e63c719b71263c4ab719492762.dmg

File: libraries/render-utils/src/HighlightEffect.cpp [hunk 1]
```diff
@@ -447,7 +470,7 @@ void SelectionToHighlight::run(const render::RenderContextPointer& renderContext
                 outputs.emplace_back(selectionName);
             }
         }
-    }
+    }*/
```
Comment (AndrewMeadows) [inline libraries/render-utils/src/HighlightEffect.cpp]: Is this commented out block cruft?  Or are we saving it for later?

File: libraries/render/src/render/HighlightStyle.h [hunk 2]
```diff
@@ -20,17 +20,24 @@ namespace render {
...
     public:
+        struct RGBA {
+            glm::vec3 color{ 1.f, 0.7f, 0.2f };
+            float alpha{ 0.9f };
+
+            RGBA(const glm::vec3& c, float a) : color(c), alpha(a) {}
+        };
+
+        RGBA _outlineUnoccluded{ { 1.f, 0.7f, 0.2f }, 0.9f };
+        RGBA _outlineOccluded{ { 1.f, 0.7f, 0.2f }, 0.9f };
```
Comment (AndrewMeadows) [inline libraries/render/src/render/HighlightStyle.h]: `1.f` --> `1.0f`

**Comment by YuppyPlays** [comment]
Testing as requested.

File: libraries/render/src/render/HighlightStyle.h [hunk 3]
```diff
@@ -20,17 +20,24 @@ namespace render {
     // This holds the configuration for a particular outline style
     class HighlightStyle {
     public:
+        struct RGBA {
+            glm::vec3 color{ 1.f, 0.7f, 0.2f };
```
Comment (AndrewMeadows) [inline libraries/render/src/render/HighlightStyle.h]: `1.f` --> `1.0f`

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/5569/HighFidelity-Beta-PR11952-63fb9b5503ff38e3ed4fcdd06a0080bf181a27e0.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/5569/HighFidelity-Beta-PR11952-63fb9b5503ff38e3ed4fcdd06a0080bf181a27e0.exe

**Pull request merged**